docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,305 @@
+
+For architectural details of where else the code could exist, the controller section makes the most sense since this proposal will be constantly modifying the state of our objects, but theoretically it could be largely put into hack with shell scripts and then a controller could simply be ordered to trigger that shell script, but this is less maintainable in the long run and not as preferred. 
+
+## Upgrade Strategy
+
+Turning off or on the feature for a particular operator is done with them setting an environment variable to enable or disable it with clusterctl initialization. They will also be able to alter their AzureMachinePool and AzureMachineTemplate instances to add the feature or remove it and that is all that is required to keep previous behavior or make use of the enhancement. No backwards compatibility will be broken, all this feature request will do is change previous controllers and add optional fields to AzureMachinePool and AzureMachineTemplate which can be utilized or not as desired.
+
+## Additional Details
+
+### Test Plan [optional]
+
+There will be e2e tests, at least one of which will be as follows:
+Have an example node and an example patch, apply the patch to the preexisting node, and then trigger the controller to pretend the interval of time has passed, and then it should attempt to create another node and compare the OS image of the new node and the original node, finding that they are both the same image.
+
+It should be tested primarily in isolation as other components shouldn’t affect what it tries to do, but it may need to be checked with other components to see what happens if certain race conditions or updates at the same time of AzureMachineTemplate are occurring (in which case a lower priority should likely be assigned to this controller for finishing its task after as ideally those changes are in effect before isolating the node).
+
+### Graduation Criteria [optional]
+
+Alpha:
+
+The feature is initially constructed and toggleable with an environment variable
+
+Beta:
+
+The feature has e2e tests implemented and is more integrated and seamless with the project.
+
+Stable: 
+
+The feature has been used for a while and is widely acceptable as well as reliable and will now be enabled by default.
+
+For AzureMachineTemplate we will move the feature enabled specification from metadata.annotations to a new optional field under spec.template.spec called nodeCachingMode which will be enabled if present and a sub field under that called cacheInterval which will be set to 1 day by default, and it will be a required subfield of nodeCachingMode if that field is there. For AzureMachinePool we will do the same thing under spec.template in the AzureMachinePool files.
```

Comment:
Have we considered defining the feature in the API from the start instead of using annotations but putting it behind a feature flag such that a validation error is returned if the field is set but the feature flag isn't enabled? E.g. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_webhook.go#L76

Reply:
I changed it to follow this style